### PR TITLE
Feature/unr 4171 add in editor multiworker toggle

### PR DESCRIPTION
**Pull request by Vix2021**
#### Description
**Add In-Editor multi worker toggle**

Users can toggle between their world’s selected LBS, and the default single server strategy without modifying any uassets when running locally. 

Users can enable/disable multi-worker from the "Start Deployment" drop-down menu in the editor or in the Editor Settings (as shown in the images below)

The Enable Multi-Worker toggle in World Settings and the command line option "-OverrideMultiWorker" have been removed as they are now redundant.

image

image

#### Release note
- You can now enable/disable the multi-worker load balancing strategy with an in-editor toggle so that no Uasset files are changed. Select `Enable Multi-Worker` from the drop option from the `Start Deployment` button on the toolbar to use the multi-worker strategy or de-select to use a single worker strategy in the editor. The `Enable Multi-Worker` toggle in World Settings and the command line option  `-OverrideMultiWorker` have been removed as they are now redundant.

#### Tests
Test cases are here: https://docs.google.com/spreadsheets/d/1TxFswDTvJYwVC6WZ4RYP92sjuYJKicNDiFWdPB98rJo/edit#gid=0

#### Documentation
Will update the zoning document on merge.

#### Known Issues
Multi-Worker class settings are no longer disabled when using the "Enable Multi-Worker" toggle.
It does not appear to be possible to disable these settings correctly unless based on another UProperty within the class. This previously relied upon bEnableMultiworker which has been removed.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp [hunk 1]
```diff
@@ -5,9 +5,12 @@
 #include "SpatialGDKSettings.h"
 #include "Utils/SpatialDebugger.h"
 
+#if WITH_EDITOR
+bool ASpatialWorldSettings::bDisableMultiWorker = false;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:9]: I'm not sure we need this variable - can't we just access the GDKEditorSettings::bDisableMultiWorker directly?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp [hunk 2]
```diff
@@ -42,6 +40,11 @@ TSubclassOf<USpatialMultiWorkerSettings> ASpatialWorldSettings::GetMultiWorkerSe
 		// behaviour).
 		return USpatialMultiWorkerSettings::StaticClass();
 	}
+	else if (!IsMultiWorkerEnabled())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:43]: Why was this moved around?

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h [hunk 3]
```diff
@@ -367,6 +367,10 @@ class SPATIALGDKEDITOR_API USpatialGDKEditorSettings : public UObject
 	UPROPERTY(EditAnywhere, config, Category = "Debug", meta = (DisplayName = "Enable spatial debugger in editor"))
 	bool bSpatialDebuggerEditorEnabled;
 
+	/** Override multi worker load balancing strategy and use single worker in the editor. */
+	UPROPERTY(EditAnywhere, config, Category = "Debug", meta = (DisplayName = "Disable multi-worker in editor"))
+	bool bDisableMultiWorker;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h:372]: I think this should be `bEnableMultiWorker`, defaults to true, and is set to false when disabling mw.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp (hunk 1 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:9]: I tried but I was unable to include SpatialGDKEditorSettings.h in WorldSettings.cpp.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp (hunk 2 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:43]: Previously the mw override was enforced locally and in the cloud. Now the override is only for the editor we want the cloud build to be unaffected so I had to change the priority of the checks to ensure the local and cloud configs were written correctly.

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h:372]: @yumnuska interested on your perspective here

Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h:372]: I agree, mainly because I generally prefer affirmative names for booleans in general as otherwise they can lead to double negative logic.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 4]
```diff
@@ -138,7 +138,6 @@ void USpatialGDKSettings::PostInitProperties()
 	// Check any command line overrides for using QBI, Offloading (after reading the config value):
 	const TCHAR* CommandLine = FCommandLine::Get();
 	CheckCmdLineOverrideBool(CommandLine, TEXT("OverrideHandover"), TEXT("Handover"), bEnableHandover);
-	CheckCmdLineOverrideOptionalBool(CommandLine, TEXT("OverrideMultiWorker"), TEXT("Multi-Worker"), bOverrideMultiWorker);
```
Comment (yumnuska) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:141]: I don't think we use this currently, but I believe we have in the past and it does concern me slightly to lose this lever. @cmsmithio do you have any thoughts?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:141]: This same functionality is now achievable via the `OverrideMultiWorkerSettingsClass` cmd line option

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h:372]: Can't deal with no double negatives

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp (hunk 1 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:9]: Discussed offline, we'll put this in the SpatialGDKSettings instead, but only use it when running in editor.

**Review by yumnuska — COMMENTED** [review]
Overall good with the spirit of this.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp (hunk 4 above)
Comment (cmsmithio) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:141]: Yea as long as we can still force single worker from the command line I think we should be good.

File: SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h (hunk 3 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDKEditor/Public/SpatialGDKEditorSettings.h]: Renamed to bEnableMultiWorker.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp (hunk 1 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp]: Moved to SpatialGDKSettings.

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h [hunk 5]
```diff
@@ -79,9 +79,11 @@ class FSpatialGDKEditorToolbarModule : public IModuleInterface, public FTickable
 	void DestroySpatialDebuggerEditor();
 	void InitialiseSpatialDebuggerEditor(UWorld* World);
 	bool IsSpatialDebuggerEditorEnabled() const;
+	bool IsMultiWorkerEditorDisabled() const;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h:82]: Would generally pref us to go with Enabled rather than Disabled getters for consistency

**Review by alastairdglennie — APPROVED** [review]
Generally LGTM, do we need a way to still disable multi-worker from CLI for tests? Or do we just pass a multi-worker settings class which just uses a single server?

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h (hunk 5 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h:82]: I am using this function for the toolbar menu option to display the checkbox and as the menu option is "Multi-Worker Disable" I need this to be a disable rather than enable check.
```
	InPluginCommands->MapAction(FSpatialGDKEditorToolbarCommands::Get().ToggleMultiWorkerEditor,
								FExecuteAction::CreateRaw(this, &FSpatialGDKEditorToolbarModule::ToggleMultiworkerEditor),
								FCanExecuteAction::CreateRaw(this, &FSpatialGDKEditorToolbarModule::OnIsSpatialNetworkingEnabled),
								FIsActionChecked::CreateRaw(this, &FSpatialGDKEditorToolbarModule::IsMultiWorkerEditorDisabled));
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp [hunk 6]
```diff
@@ -42,6 +36,11 @@ TSubclassOf<USpatialMultiWorkerSettings> ASpatialWorldSettings::GetMultiWorkerSe
...
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp:39]: I don't think we need `ASpatialWorldSettings::IsMultiWorkerEnabled` anymore. It only returns false if the editor config has been set, so we should remove this function, and put it into `USpatialStatics::IsSpatialMultiWorkerEnabled`

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 7]
```diff
@@ -281,4 +281,11 @@ bool USpatialGDKSettings::GetPreventClientCloudDeploymentAutoConnect() const
 	return (IsRunningGame() || IsRunningClientOnly()) && bPreventClientCloudDeploymentAutoConnect;
 };
 
+#if WITH_EDITOR
+void USpatialGDKSettings::SetMultiWorkerEditor(bool IsEnabled)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:285]:
```suggestion
void USpatialGDKSettings::SetMultiWorkerEnabled(bool bIsEnabled)
```

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 8]
```diff
@@ -281,4 +281,11 @@ bool USpatialGDKSettings::GetPreventClientCloudDeploymentAutoConnect() const
...
+{
+	bEnableMultiWorker = IsEnabled;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp:287]:
```suggestion
	bEnableMultiWorker = bIsEnabled;
```

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 9]
```diff
@@ -248,13 +248,22 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
 	UPROPERTY(EditAnywhere, config, Category = "Debug", meta = (MetaClass = "SpatialDebugger"))
 	TSubclassOf<ASpatialDebugger> SpatialDebugger;
 
+	/** Enables multi-worker, if not checked uses single worker strategy in the editor.  */
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h:251]:
```suggestion
	/** Enables multi-worker, if false uses single worker strategy in the editor.  */
```

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h:82]: I think we can just make the toggle `Enable Multi-Worker`.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp (hunk 6 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialWorldSettings.cpp]: Of course, that is a neater solution.

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h (hunk 5 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDKEditorToolbar/Public/SpatialGDKEditorToolbar.h:82]: Sure I will keep it all consistent with enable everywhere.

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 10]
```diff
@@ -248,9 +248,17 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
 	UPROPERTY(EditAnywhere, config, Category = "Debug", meta = (MetaClass = "SpatialDebugger"))
 	TSubclassOf<ASpatialDebugger> SpatialDebugger;
 
+	/** Enables multi-worker, if false uses single worker strategy in the editor.  */
+	UPROPERTY(EditAnywhere, config, Category = "Debug", meta = (DisplayName = "Enable multi-worker in editor"))
+	bool bEnableMultiWorker;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h:253]:
```suggestion
	bool bEnableMultiWorker;
	
```

**Review by m-samiec — APPROVED** [review]
Nice!
